mindsdb/integrations/libs/ml_handler_proc.py

```diff
@@ -0,0 +1,227 @@
+"""
+Utility functions for run ml handler in subprocess.
+Interation is performed using stdin, stdout.
+
+Flow
+    1. subprocess starts and initialize ml hanlder instance inside itself
+    2. waits incoming command from stdin
+    3. passes this command to ml handler instance
+    4. sends response from ml handler to stdout
+
+Protocol over stdin/out:
+
+- [8 bytes] length of the command
+- [length] command: is picked dict with keys: method, args, kwargs
+
+if 'error' in response dict, parent process trows exception
+
+"""
+
+import pickle
+import subprocess
+import sys
+import struct
+import traceback
+import importlib
+import threading
+import queue
+from threading import Lock
+
+
+# =================  child process ====================
+
+class MLHandlerProcess:
+
+    def __init__(self):
+        self.ml_instance = None
+
+    def mainloop(self):
+        # replace print output to stderr
+        sys.stdout = sys.stderr
+
+        self.stdin = open(0, 'rb')
+        self.stdout = open(1, 'wb')
+
+        while True:
+            params = self.read_input()
+            try:
+                method = params['method']
+                args = params.get('args', [])
+                kwargs = params.get('kwargs', {})
+
+                if hasattr(self, method):
+                    # it is control method
+                    ret = getattr(self, method)(*args, **kwargs)
+                else:
+                    # it is handler method
+                    ret = getattr(self.ml_instance, method)(*args, **kwargs)
+
+            except SystemExit:
+                raise
+            except:
+                error = traceback.format_exc()
+                self.send_output({'error': error})
+                continue
+
+            self.send_output(ret)
+
+    def send_output(self, obj):
+        # read stdin
+        encoded = pickle.dumps(obj)
+
+        length_enc = struct.pack('L', len(encoded))
+
+        self.stdout.write(length_enc)
+        self.stdout.write(encoded)
+        self.stdout.flush()
+
+    def read_input(self):
+        # write to stdout
+        length_enc = self.stdin.read(8)
+        length = struct.unpack('L', length_enc)[0]
+
+        encoded = self.stdin.read(length)
+        obj = pickle.loads(encoded)
+
+        return obj
+
+    def init_handler(self, class_path, company_id, integration_id, predictor_id):
+        # mdb initialization
+        import mindsdb.interfaces.storage.db as db
+        db.init()
+
+        from mindsdb.interfaces.storage.model_fs import ModelStorage, HandlerStorage
+
+        module_name, class_name = class_path
+        module = importlib.import_module(module_name)
+        klass = getattr(module, class_name)
```

Comment:
Just a nitpick: I thought the standard alternate naming for "class" in python was "cls", though I can't find a source right now.

Reply:
Max renamed it in ml_exec_base.py
I will rename it the same